File: cypress_ai_reporter/lib/model_training/predict.py
import tensorflow as tf
import numpy as np
import os
from PIL import Image

# ✅ Load the trained model
model = tf.keras.models.load_model("inceptionv3_model.h5")
print("✅ Model Loaded Successfully!")

# ✅ Function to preprocess image
def preprocess_image(image_path, target_size=(150, 150)):
    img = Image.open(image_path).convert("RGB")  # Ensure 3 channels (RGB)
    img = img.resize(target_size)  # Resize to model input shape
    img_array = np.array(img) / 255.0  # Normalize (0-1)
    img_array = np.expand_dims(img_array, axis=0)  # Add batch dimension
    return img_array

# ✅ Function to predict image class
def predict_image(image_path):
    img_array = preprocess_image(image_path)
    prediction = model.predict(img_array)
    predicted_class = np.argmax(prediction)  # Get class with highest probability
    confidence = np.max(prediction)  # Get confidence score
    return predicted_class, confidence

# ✅ Predict for all images in a directory
def predict_images_from_directory(directory):
    image_files = [f for f in os.listdir(directory) if f.lower().endswith(('.jpg', '.jpeg', '.png'))]
    
    for file in image_files:
        image_path = os.path.join(directory, file)
        predicted_class, confidence = predict_image(image_path)
        print(f"🖼️ {file} → Class {predicted_class}, Confidence: {confidence:.4f}")

# ✅ Run predictions on test images
predict_images_from_directory("frames/images")  # Change to your image directory

File: cypress_ai_reporter/lib/model_training/predict.js
const tf = require('@tensorflow/tfjs-node');
const fs = require('fs');
const path = require('path');
const { createCanvas, loadImage } = require('canvas');

// ✅ Load TensorFlow.js model
async function loadModel() {
    console.log("🔄 Loading model...");
    const model = await tf.loadLayersModel('file://tfjs_model/model.json');
    console.log("✅ Model loaded successfully!");
    return model;
}

// ✅ Preprocess image (resize, normalize)
async function preprocessImage(imagePath) {
    const img = await loadImage(imagePath);
    const canvas = createCanvas(150, 150);
    const ctx = canvas.getContext('2d');
    ctx.drawImage(img, 0, 0, 150, 150);

    const imageData = ctx.getImageData(0, 0, 150, 150);
    let tensor = tf.browser.fromPixels(imageData)
        .toFloat()
        .expandDims(0);  // Add batch dimension
    tensor = tensor.div(tf.scalar(255));  // Normalize
    return tensor;
}

// ✅ Predict Class for Each Image
async function predictImages(dirPath) {
    const model = await loadModel();
    const imageFiles = fs.readdirSync(dirPath).filter(file => /\.(jpg|jpeg|png)$/i.test(file));

    for (const file of imageFiles) {
        const imagePath = path.join(dirPath, file);
        const tensor = await preprocessImage(imagePath);
        const prediction = model.predict(tensor);
        const probabilities = await prediction.data();
        const predictedClass = probabilities.indexOf(Math.max(...probabilities));

        console.log(`🖼️ ${file} → Class ${predictedClass}, Confidence: ${Math.max(...probabilities).toFixed(4)}`);
    }
}

// ✅ Run Prediction
predictImages('./images/test');  // Change directory as needed
